Mathlib/LinearAlgebra/Matrix/PosDef.lean

```diff
@@ -38,16 +43,36 @@ open scoped Matrix
 -/
 
 /-- A matrix `M : Matrix n n R` is positive semidefinite if it is hermitian
-   and `xᴴMx` is nonnegative for all `x`. -/
+   and `xᴴ M x` is nonnegative for all `x`. -/
 def PosSemidef (M : Matrix n n R) :=
   M.IsHermitian ∧ ∀ x : n → R, 0 ≤ dotProduct (star x) (M.mulVec x)
 #align matrix.pos_semidef Matrix.PosSemidef
 
-theorem PosSemidef.re_dotProduct_nonneg {M : Matrix n n 𝕜} (hM : M.PosSemidef) (x : n → 𝕜) :
+/-- A diagonal matrix is positive semidefinite iff its diagonal entries are nonnegative. -/
+lemma posSemidef_diagonal_iff_nonneg [DecidableEq n] {d : n → R} :
+    PosSemidef (diagonal d) ↔ (∀ i : n, 0 ≤ d i) := by
+  constructor
+  · intro ⟨_, hP⟩ i
+    simpa using hP (Pi.single i 1)
+  · intro hd
+    constructor
+    · rw [IsHermitian, conjTranspose, diagonal_transpose, diagonal_map (star_zero R)]
+      exact congr_arg _ (funext fun i ↦ star_eq_self_of_nonneg (hd i))
+    · intro x
+      rw [dotProduct]
+      refine Finset.sum_nonneg fun i _ ↦ ?_
+      simpa only [mulVec_diagonal, ← mul_assoc] using conjugate_nonneg (hd i) _
+
+namespace PosSemidef
+
+theorem isHermitian {M : Matrix n n R} (hM : M.PosSemidef) : M.IsHermitian :=
+  hM.1
+
+theorem re_dotProduct_nonneg {M : Matrix n n 𝕜} (hM : M.PosSemidef) (x : n → 𝕜) :
     0 ≤ IsROrC.re (dotProduct (star x) (M.mulVec x)) :=
   IsROrC.nonneg_iff.mp (hM.2 _) |>.1
 
-theorem PosSemidef.submatrix {M : Matrix n n R} (hM : M.PosSemidef) (e : m ≃ n) :
+theorem submatrix {M : Matrix n n R} (hM : M.PosSemidef) (e : m ≃ n) :
```

Comment:
why is it required that `e` is an equiv here? (I understand this lemma is not yours, but if this is easy to fix by replacing it with a general map, it might be worth doing it).

Reply:
The lemma is true as stated for any e (it needn't be injective or surjective -- you need injectivity for the positive-definite version, but not for positive semidef). Mathematically this is just the special case of `conjTranspose_mul_mul_same` when `B` is a matrix of 0's and 1's (with `B i j = 1` iff `e j = i`). However, the proof would need to be completely rewritten; and doing it properly probably require developing some more API about how `mulVec` and `vecMul` interact with `Submatrix`. 

Can I suggest that this be pursued separately from this PR?

Reply:
PS. It turns out that the necessary API already exists (specifically `Matrix.submatrix_mul`), so I wrote a new proof for the general case which is shorter than the existing one for equivs.